Type guards not working when using partial mapped types.

<!-- 🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨

Half of all issues filed here are duplicates, answered in the FAQ, or not appropriate for the bug tracker. Even if you think you've found a *bug*, please read the FAQ first, especially the Common "Bugs" That Aren't Bugs section!

Please help us by doing the following steps before logging an issue:
  * Search: https://github.com/Microsoft/TypeScript/search?type=Issues
  * Read the FAQ: https://github.com/Microsoft/TypeScript/wiki/FAQ

Please fill in the *entire* template below.
-->

<!-- Please try to reproduce the issue with `typescript@next`. It may have already been fixed. -->
**TypeScript Version:** 3.1.3

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:**
Type Guards, Partial

**Code**

```ts
type Obj = {} | undefined;

type User = {
  email: string;
  name: string;
};

type PartialUser = Partial<User>;

// type PartialUser = {
//   email?: string;
//   name?: string;
// };

function isUser(obj: Obj): obj is PartialUser {
  return true;
}

function getUserName(obj: Obj) {
  if (isUser(obj)) {
    return obj.name;
  }

  return '';
}
```

**Expected behavior:**
Successful compilation of the above code with `tsc --strictNullChecks`

**Actual behavior:**
Compilation fails with `error TS2339: Property 'name' does not exist on type '{}'`.
Interestingly, replacing `type PartialUser = Partial<User>;` with

```
type PartialUser = {
  email?: string;
  name?: string;
};
```

works as expected. Shouldn't these two definitions of `PartialUser` be functionally equivalent?

**Playground Link:** <!-- A link to a TypeScript Playground "Share" link which demonstrates this behavior -->
https://agentcooper.github.io/typescript-play/?noImplicitAny=false&strictFunctionTypes=false&strictPropertyInitialization=false&noImplicitThis=false&noImplicitReturns=false&alwaysStrict=false#code/C4TwDgpgBA8gRgKygXigbwL5QD5QK4B2AJhAGYCWBERA3AFB2iRQCqAzhAE4rp1T9QIAWwCG5ADYAuKG2CdKAc3oCoBEUIjTZ8gkroZ6jcNAAKIzsHIjx7LjzMWr4gDy3OAPkMB6L1CanzS2s3HjQ6HwFhMXEAfi05RXoI-jUNOJkE3STfAwZSQgBjSwB7AihyNjcACmLEaXgEAEppWqQKqAcgmw5uMP5OCGA8TjK5PAh6DDzCkrKFQbcAOXUIGrrYREbefnJSKCqK6tbGrb6BAaGRqFaAOlSJvigpx4vhsoByd8mgA

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -12567,7 +12567,7 @@ namespace ts {
                 }
                 else {
                     // An empty object type is related to any mapped type that includes a '?' modifier.
-                    if (isPartialMappedType(target) && isEmptyObjectType(source)) {
+                    if (relation !== subtypeRelation && isPartialMappedType(target) && isEmptyObjectType(source)) {
                         return Ternary.True;
                     }
                     if (isGenericMappedType(target)) {

--- a/tests/baselines/reference/partialTypeNarrowedToByTypeGuard.js
+++ b/tests/baselines/reference/partialTypeNarrowedToByTypeGuard.js
@@ -1,0 +1,42 @@
+//// [partialTypeNarrowedToByTypeGuard.ts]
+type Obj = {} | undefined;
+
+type User = {
+    email: string;
+    name: string;
+};
+
+type PartialUser = Partial<User>;
+
+// type PartialUser = {
+//   email?: string;
+//   name?: string;
+// };
+
+function isUser(obj: Obj): obj is PartialUser {
+    return true;
+}
+
+function getUserName(obj: Obj) {
+    if (isUser(obj)) {
+        return obj.name;
+    }
+
+    return '';
+}
+
+//// [partialTypeNarrowedToByTypeGuard.js]
+"use strict";
+// type PartialUser = {
+//   email?: string;
+//   name?: string;
+// };
+function isUser(obj) {
+    return true;
+}
+function getUserName(obj) {
+    if (isUser(obj)) {
+        return obj.name;
+    }
+    return '';
+}

--- a/tests/baselines/reference/partialTypeNarrowedToByTypeGuard.symbols
+++ b/tests/baselines/reference/partialTypeNarrowedToByTypeGuard.symbols
@@ -1,0 +1,52 @@
+=== tests/cases/compiler/partialTypeNarrowedToByTypeGuard.ts ===
+type Obj = {} | undefined;
+>Obj : Symbol(Obj, Decl(partialTypeNarrowedToByTypeGuard.ts, 0, 0))
+
+type User = {
+>User : Symbol(User, Decl(partialTypeNarrowedToByTypeGuard.ts, 0, 26))
+
+    email: string;
+>email : Symbol(email, Decl(partialTypeNarrowedToByTypeGuard.ts, 2, 13))
+
+    name: string;
+>name : Symbol(name, Decl(partialTypeNarrowedToByTypeGuard.ts, 3, 18))
+
+};
+
+type PartialUser = Partial<User>;
+>PartialUser : Symbol(PartialUser, Decl(partialTypeNarrowedToByTypeGuard.ts, 5, 2))
+>Partial : Symbol(Partial, Decl(lib.es5.d.ts, --, --))
+>User : Symbol(User, Decl(partialTypeNarrowedToByTypeGuard.ts, 0, 26))
+
+// type PartialUser = {
+//   email?: string;
+//   name?: string;
+// };
+
+function isUser(obj: Obj): obj is PartialUser {
+>isUser : Symbol(isUser, Decl(partialTypeNarrowedToByTypeGuard.ts, 7, 33))
+>obj : Symbol(obj, Decl(partialTypeNarrowedToByTypeGuard.ts, 14, 16))
+>Obj : Symbol(Obj, Decl(partialTypeNarrowedToByTypeGuard.ts, 0, 0))
+>obj : Symbol(obj, Decl(partialTypeNarrowedToByTypeGuard.ts, 14, 16))
+>PartialUser : Symbol(PartialUser, Decl(partialTypeNarrowedToByTypeGuard.ts, 5, 2))
+
+    return true;
+}
+
+function getUserName(obj: Obj) {
+>getUserName : Symbol(getUserName, Decl(partialTypeNarrowedToByTypeGuard.ts, 16, 1))
+>obj : Symbol(obj, Decl(partialTypeNarrowedToByTypeGuard.ts, 18, 21))
+>Obj : Symbol(Obj, Decl(partialTypeNarrowedToByTypeGuard.ts, 0, 0))
+
+    if (isUser(obj)) {
+>isUser : Symbol(isUser, Decl(partialTypeNarrowedToByTypeGuard.ts, 7, 33))
+>obj : Symbol(obj, Decl(partialTypeNarrowedToByTypeGuard.ts, 18, 21))
+
+        return obj.name;
+>obj.name : Symbol(name, Decl(partialTypeNarrowedToByTypeGuard.ts, 3, 18))
+>obj : Symbol(obj, Decl(partialTypeNarrowedToByTypeGuard.ts, 18, 21))
+>name : Symbol(name, Decl(partialTypeNarrowedToByTypeGuard.ts, 3, 18))
+    }
+
+    return '';
+}

--- a/tests/baselines/reference/partialTypeNarrowedToByTypeGuard.types
+++ b/tests/baselines/reference/partialTypeNarrowedToByTypeGuard.types
@@ -1,0 +1,49 @@
+=== tests/cases/compiler/partialTypeNarrowedToByTypeGuard.ts ===
+type Obj = {} | undefined;
+>Obj : Obj
+
+type User = {
+>User : User
+
+    email: string;
+>email : string
+
+    name: string;
+>name : string
+
+};
+
+type PartialUser = Partial<User>;
+>PartialUser : Partial<User>
+
+// type PartialUser = {
+//   email?: string;
+//   name?: string;
+// };
+
+function isUser(obj: Obj): obj is PartialUser {
+>isUser : (obj: Obj) => obj is Partial<User>
+>obj : Obj
+
+    return true;
+>true : true
+}
+
+function getUserName(obj: Obj) {
+>getUserName : (obj: Obj) => string | undefined
+>obj : Obj
+
+    if (isUser(obj)) {
+>isUser(obj) : boolean
+>isUser : (obj: Obj) => obj is Partial<User>
+>obj : Obj
+
+        return obj.name;
+>obj.name : string | undefined
+>obj : Partial<User>
+>name : string | undefined
+    }
+
+    return '';
+>'' : ""
+}

--- a/tests/cases/compiler/partialTypeNarrowedToByTypeGuard.ts
+++ b/tests/cases/compiler/partialTypeNarrowedToByTypeGuard.ts
@@ -1,0 +1,26 @@
+// @strict: true
+type Obj = {} | undefined;
+
+type User = {
+    email: string;
+    name: string;
+};
+
+type PartialUser = Partial<User>;
+
+// type PartialUser = {
+//   email?: string;
+//   name?: string;
+// };
+
+function isUser(obj: Obj): obj is PartialUser {
+    return true;
+}
+
+function getUserName(obj: Obj) {
+    if (isUser(obj)) {
+        return obj.name;
+    }
+
+    return '';
+}

